include/deal.II/lac/sparse_direct.h

```diff
@@ -452,97 +452,202 @@ class SparseDirectUMFPACK : public EnableObserverPointer
+  initialize(const Matrix &matrix);
 
   /**
-   * Copy the computed solution into the solution vector.
+   * A function in which the inverse of the matrix is applied to the input
+   * vector <tt>src</tt> and the solution is written into the output vector
+   * <tt>dst</tt>.
    */
   void
-  copy_solution(Vector<double> &vector);
+  vmult(Vector<double> &dst, const Vector<double> &src) const;
 
   /**
-   *
+   * A function in which the inverse of the transposed matrix is applied to the
+   * input vector <tt>src</tt> and the solution is written into the output
+   * vector <tt>dst</tt>.
    */
   void
-  copy_rhs_to_mumps(const Vector<double> &rhs);
+  Tvmult(Vector<double> &dst, const Vector<double> &src) const;
 
   /**
-   * Flags storing whether the function <tt>initialize ()</tt> has already
-   * been called.
+   * A function that returns the ICNTL integer array from MUMPS.
+   * The ICNTL array contains integer control parameters for the MUMPS solver.
+   * Keep in mind that MUMPS is a fortran library and the documentation refers
+   * to indices into this array into this array starting from one rather than
+   * from zero. To select the correct index one can use a macro like this:
+   * `#define ICNTL(I) icntl[(I)-1]`. In the MUMPS documentation there is the
+   * description of each parameter of the array. Be aware that ownership of the
+   * array remains in the current class rather than with the caller of this
+   * function.
    */
-  bool initialize_called;
+  int *
+  get_icntl();
+
+private:
+#ifdef DEAL_II_WITH_MUMPS
+  mutable DMUMPS_STRUC_C id;
+#endif // DEAL_II_WITH_MUMPS
 
-public:
   /**
-   * Declare type for container size.
+   * a contains the actual values of the matrix. a[k] is the value of the matrix
+   * entry (i,j) if irn[k] == i and jcn[k] == j.
    */
-  using size_type = types::global_dof_index;
+  double *a;
 
   /**
-   * Constructor
+   * The right-hand side vector. This is the right hand side of the system
    */
-  SparseDirectMUMPS();
+  mutable std::vector<double> rhs;
 
   /**
-   * Destructor
+   * irn contains the row indices of the non-zero entries of the matrix.
    */
-  ~SparseDirectMUMPS();
+  int *irn;
 
   /**
-   * Exception
+   * jcn contains the column indices of the non-zero entries of the matrix.
    */
-  DeclException0(ExcInitializeAlreadyCalled);
+  int *jcn;
 
   /**
-   * This function initializes a MUMPS instance and hands over the system's
-   * matrix <tt>matrix</tt> and right-hand side <tt>rhs_vector</tt> to the
-   * solver.
+   * The number of rows of the matrix. The matrix is square.
    */
-  template <class Matrix>
-  void
-  initialize(const Matrix &matrix, const Vector<double> &rhs_vector);
+  types::global_dof_index n;
 
   /**
-   * This function initializes a MUMPS instance and computes the factorization
-   * of the system's matrix <tt>matrix</tt>.
+   * The number of non-zero entries in the matrix.
+   */
+  types::global_dof_index nz;
```

Comment:
People generally write this as `nnz` (n_nonzeros). Separately, though, I can relatively easily imagine matrices where you have a few million rows, but more than 4B nonzeros (at least in parallel). Is this a variable you have to pass to MUMPS at some point? Otherwise, I think you want to use `std::size_t` for this variable, or perhaps `std::uint64_t`.

Perhaps seen a different way: It makes sense to say that `n` is of type `types::global_dof_index` because it represents the total number of DoFs. But the total number of nonzeros is unrelated to the number of DoFs, and so it makes sense to also expect that it would have a different type.